docs/architecture/adr-049-state-sync-hooks.md

```diff
@@ -0,0 +1,187 @@
+# ADR 049: State Sync Hooks
+
+## Changelog
+
+- Jan 19, 2022: Initial Draft
+
+## Status
+
+Draft, Under Implementation
+
+## Abstract
+
+This ADR provides hooks for app modules to publish snapshot of additional state(outside of IAVL tree) for state-sync.
+
+## Context
+
+New clients use state-sync to download snapshots of module state from peer nodes. Currently, the snapshot consists of a
+stream of `SnapshotStoreItem` and `SnapshotIAVLItem`, which means for app modules that maintain their states outside of
+the IAVL tree, they can not add their states to the snapshot stream for state-sync.
+
+Notification: `Even though the module state data is outside of the tree, for determinism we
+require that the hash of the external data should be posted in the IAVL tree`
+
+## Decision
+
+A simple proposal based on our existing implementation is that, we can add two new message types: `SnapshotExtensionMeta` 
+and `SnapshotExtensionPayload`, and they are appended to the existing multi-store stream with `SnapshotExtensionMeta` 
+acting as a delimiter between extensions. As the chunk hashes should be able to ensure data integrity, we don't need 
+a delimiter to mark the end of the snapshot stream.
+
+Besides, we provide `Snapshotter` and `ExtensionSnapshotter` interface for modules to implement snapshotters, which will handle both taking 
+snapshot and the restoration. Each module could have mutiple snapshotters, and for modules with additional state, they should
+implement `ExtensionSnapshotter` as extension snapshotters. When setting up the application, the snapshot `Manager` should call 
+`RegisterExtensions([]ExtensionSnapshotter…)` to register all the extension snapshotters.
+
+```proto
+// SnapshotItem is an item contained in a rootmulti.Store snapshot.
+message SnapshotItem {
+  // item is the specific type of snapshot item.
+  oneof item {
+    SnapshotStoreItem        store             = 1;
+    SnapshotIAVLItem         iavl              = 2 [(gogoproto.customname) = "IAVL"];
+    SnapshotExtensionMeta    extension         = 3;
+    SnapshotExtensionPayload extension_payload = 4;
+  }
+}
+
+// SnapshotStoreItem contains metadata about a snapshotted store.
+message SnapshotStoreItem {
+  string name = 1;
+}
+
+// SnapshotIAVLItem is an exported IAVL node.
+message SnapshotIAVLItem {
+  bytes key     = 1;
+  bytes value   = 2;
+  int64 version = 3;
+  int32 height  = 4;
+}
```

Comment:
If we're not modifying existing types, I don't think we need to include them here for brevity.

Reply:
ok , removed `SnapshotStoreItem` and `SnapshotIAVLItem`